### PR TITLE
restrict max concurrent outbound dials

### LINKED ISSUE #6778: max_concurrent_outbound_dials does not restrict max concurrent outbound dials
### Describe the feature

The `max_concurrent_outbound_dials` 
https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L832-L834

only restricts how many __new__ outbound dials we create __at once__ but this does not respect how many are already in progress atm.

https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L702-L702

This should instead restrict how many outbound dials we allow at once and should slow down new connections if exceeded.

## TODO

This can be solved by a new connection state and counter that is `PendingOut`

https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L822-L824

https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L1021-L1022

when we initiate an outgoing connection the peer's state is set to `PendingOut` and counter increased.

once the session is established or failed we need to update the state and counter for all possible cases:

https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L850-L850

https://github.com/paradigmxyz/reth/blob/a7e183d1a58f3d7eada9a087eca56af5e794e916/crates/net/network/src/peers/manager.rs#L386-L386

and need a new function `on_active_session_established` that then shifts the peer's state from `PendingOut` to `Out`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by abhijeetbhagat**
closes #6778

### CODE REVIEW

File: crates/net/network/src/peers/manager.rs [hunk 1]
```diff
@@ -429,6 +429,15 @@ impl PeersManager {
         self.fill_outbound_slots();
     }
 
+    /// Called when a _pending_ outbound connection is successful.
+    pub(crate) fn on_active_session_established(&mut self, peer_id: PeerId) {
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:433]: this should include `_outgoing_`

File: crates/net/network/src/peers/manager.rs [hunk 2]
```diff
@@ -832,14 +835,17 @@ pub struct ConnectionInfo {
     /// Maximum allowed concurrent outbound dials.
     #[cfg_attr(feature = "serde", serde(default))]
     max_concurrent_outbound_dials: usize,
+    /// Counter for pending outbound connections.
+    num_pendingout: usize,
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:839]: please move to the other counters and add serde(skip)

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks pretty good!
I believe this is already complete.

I'd like to see another test case that adds a few peers
* starts all allowed pendingOut (until max_concurrent is satisfied)
* calls on_active_outgoing_established 
ensures peers are now `::Out` and counters are as expected

**Comment by abhijeetbhagat** [comment]
@mattsse can you please check if the test case looks good?

File: crates/net/network/src/peers/manager.rs [hunk 3]
```diff
@@ -2283,4 +2305,45 @@ mod tests {
...
+        // establish dialed connections
+        for peer_id in num_pendingout_states.iter() {
+            peer_manager.on_active_outgoing_established(*peer_id);
+        }
+
+        // all dialed connections should now be in 'Out' state
+        for peer_id in num_pendingout_states.iter() {
+            assert_eq!(peer_manager.peers.get(peer_id).unwrap().state, PeerConnectionState::Out);
+        }
+    }
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:2348]: this looks good,
I'd like to also test the emitted events by polling the peer manager, see other tests for examples

**Review by mattsse — CHANGES_REQUESTED** [review]
last test nit

**Comment by abhijeetbhagat** [comment]
@mattsse please check.

**Comment by abhijeetbhagat** [comment]
@mattsse please check

**Review by mattsse — APPROVED** [review]
lgtm!

ty!

**Pull request merged**